Core/JsRuntime/InternalInclude/Babylon/JsRuntimeScheduler.h

```diff
@@ -1,30 +1,146 @@
+    //      when `Rundown()` is called in the N-API class destructor.
+    class JsRuntimeScheduler : public JsRuntime::IDisposingCallback
     {
     public:
         explicit JsRuntimeScheduler(JsRuntime& runtime)
-            : m_runtime{runtime}
+            : m_runtime{&runtime}
+            , m_scheduler{*this}
         {
+            std::scoped_lock lock{m_mutex};
+            JsRuntime::RegisterDisposing(*m_runtime, this);
         }
 
-        template<typename CallableT>
-        void operator()(CallableT&& callable) const
+        ~JsRuntimeScheduler()
+        {
+            assert(m_count == 0 && "Schedulers for the JavaScript thread are not yet invoked");
+
+            std::scoped_lock lock{m_mutex};
+            if (m_runtime != nullptr)
+            {
+                JsRuntime::UnregisterDisposing(*m_runtime, this);
+            }
+        }
+
+        // Wait until all of the schedulers are invoked.
+        void Rundown()
         {
-            m_runtime.Dispatch([callable{std::forward<CallableT>(callable)}](Napi::Env) {
-                callable();
-            });
+            while (m_count > 0)
+            {
+                m_disposingDispatcher.blocking_tick(arcana::cancellation::none());
+            }
+        }
+
+        // Get a scheduler to invoke continuations on the JavaScript thread.
+        const auto& Get()
+        {
+            ++m_count;
+            return m_scheduler;
         }
 
     private:
-        JsRuntime& m_runtime;
+        void Disposing() override
+        {
+            std::scoped_lock lock{m_mutex};
+            m_runtime = nullptr;
+        }
+
+        class SchedulerImpl
+        {
+        public:
+            explicit SchedulerImpl(JsRuntimeScheduler& parent)
+                : m_parent{parent}
+            {
+            }
+
+            template<typename CallableT>
+            void operator()(CallableT&& callable) const
+            {
+                m_parent.Dispatch(std::forward<CallableT>(callable));
+            }
+
+        private:
+            JsRuntimeScheduler& m_parent;
+        };
+
+        template<typename CallableT>
+        void Dispatch(CallableT&& callable)
+        {
+            std::scoped_lock lock{m_mutex};
+
+            if (m_runtime != nullptr)
+            {
+                m_runtime->Dispatch([callable{std::forward<CallableT>(callable)}](Napi::Env) {
+                    callable();
+                });
+            }
+            else
+            {
+                m_disposingDispatcher(std::forward<CallableT>(callable));
+            }
+
+            --m_count;
+        }
+
+        mutable std::mutex m_mutex{};
+        JsRuntime* m_runtime{};
+        std::function<void()> m_disposingCallback{};
+
+        SchedulerImpl m_scheduler;
+        std::atomic<int> m_count{0};
+        arcana::manual_dispatcher<128> m_disposingDispatcher{};
```

Comment:
If each `JsRuntimeScheduler` instance has its own dispatch queue, and it is only pumped by a call to `Rundown` in the destructor, then if we ever have a situation where one object is trying to call into another object (as part of the async code executing during destruction), then the separate dispatch queue in the separate object will be in a state where it is getting pumped and we will basically deadlock. It seems like this is fixable by having a shared dispatch queue across all objects, and a given object's destructor's call to Rundown is pumping for all queued work. I don't think this breaks anything, but I think it would prevent this potential deadlock situation. I'm not sure where a shared dispatch queue would exist... maybe we'd need to push some of this logic up to JSRuntime.